flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StateAssignmentOperation.java

```diff
@@ -99,32 +119,40 @@ public void assignStates() {
 						operatorID,
 						executionJobVertex.getParallelism(),
 						executionJobVertex.getMaxParallelism());
-				} else if (operatorState.getNumberCollectedStates() > 0) {
-					statelessSubTasks = false;
 				}
-				operatorStates.add(operatorState);
+				operatorStates.put(operatorIDPair.getGeneratedOperatorID(), operatorState);
 			}
-			if (!statelessSubTasks) { // skip tasks where no operator has any state
-				assignAttemptState(executionJobVertex, operatorStates);
+
+			final TaskStateAssignment stateAssignment = new TaskStateAssignment(executionJobVertex,	operatorStates);
+			vertexAssignments.put(executionJobVertex, stateAssignment);
+			for (final IntermediateResult producedDataSet : executionJobVertex.getInputs()) {
+				consumerAssignment.put(producedDataSet, stateAssignment);
```

Comment:
`UnionGate` reads two or more different `IntermediateResult`. `IntermediateResult` translates directly into subpartition/gate.
However, it's a good point that `IntermediateResult` doesn't override equals/hashCode. It currently works, because the JobGraph is a graph where the objects are connected, but I don't know if we can rely on the assumption. I'm switching to `IntermediateDataSetID`.

Reply:
I see, thanks for the clarification and update.